server/interpreter.js

```diff
@@ -4812,30 +4845,34 @@ stepFuncs_['ForInStatement'] = function (stack, state, node) {
  * @return {!Interpreter.State|undefined}
  */
 stepFuncs_['ForStatement'] = function (stack, state, node) {
-  var mode = state.mode_ || 0;
-  if (mode === 0) {
-    state.mode_ = 1;
-    if (node['init']) {
-      return new Interpreter.State(node['init'], state.scope);
-    }
-  } else if (mode === 1) {
-    state.mode_ = 2;
-    if (node['test']) {
-      return new Interpreter.State(node['test'], state.scope);
-    }
-  } else if (mode === 2) {
-    state.mode_ = 3;
-    if (node['test'] && !state.value) {
-      // Done, exit loop.
-      stack.pop();
-    } else {  // Execute the body.
-      state.isLoop = true;
-      return new Interpreter.State(node['body'], state.scope);
-    }
-  } else if (mode === 3) {
-    state.mode_ = 1;
-    if (node['update']) {
-      return new Interpreter.State(node['update'], state.scope);
+  // If we've just evaluated node.test, and result was false, terminate loop.
+  if  (state.step_ === 2 && !state.value) {
+    stack.pop();
+    return;
+  }
+  while (true) {
+    switch (state.step_) {
+      case 0:  // Eval init expression.
+        state.step_ = 1;
+        state.isLoop = true;  // TODO(cpcallen): remove or declare.
+        if (node['init']) {
+          return new Interpreter.State(node['init'], state.scope);
+        }
+        // FALL THROUGH
+      case 1: // Eval test expression.
```

Comment:
Two spaces before //
0 has this, 1-3 don't.

Reply:
Done.